### DISCUSSION TITLE
Adding routes dynamically at runtime?

### DISCUSSION THREAD

**Discussion by Jumborner** [nuxt/nuxt#9970] [Questions (Nuxt 2)]
Hello, 
I have vue project in which I would like to implement SSR, so I've decided to try out nuxt 3 and port my project to it and got stuck.

In vue, inside of main.js I make an api call with the user's ip adress which then retreives me an array with routes and the meta parameters for them, afterwards I use router.addRoute and add the routes. It is unclear to me how I can implement this in nuxt with what I've seen on its dynamic routes.

I have tried making a plugin, which would create add the routes `export default ({ app }) => { app.router.addRoute({ path: '/testroute', name: 'RouteTest', component: () => import('../views/extraRoutes/test_route.vue'), meta: { layout: 'AppLayout_3125615793803362412', }, }) }`

I'm getting addRoute function undefined. Tried by creating a routes array and adding it, still doesnt work.

If someone could point me towards the right direction would be grateful

**Comment by svenhue** [comment]
Instead of addRoutes use extendRoutes from nuxt
https://nuxtjs.org/docs/configuration-glossary/configuration-router/#extendroutes

**Comment by schellenbergk** [comment]
What about adding a new route at runtime? What if I need to call an API to get some extra routes?

**Reply by gianpesto** [reply]
in nuxt 3, you can do that by adding a router plugin that use addRoutes:
```javascript
// ~/plugins/addRoutes.js

export default defineNuxtPlugin({
  name: "addRoutes",

 setup(nuxtApp) {
      nuxtApp.$router.addRoute({
        path: '/some-weird-foo-path',
        name: 'uniqueFooName',
        component: '~/pages/foo.vue',
      });
    }
});
```

anyway, note that nuxt automatically ads `/foo` as route automatically from the `~/pages` directory. In order to prevent this you can use the [`pages:extend` hook from within your nuxt.config](https://nuxt.com/docs/guide/going-further/custom-routing#using-the-pagesextend-hook)

**Reply by gianpesto** [reply]
i have to add something to my answer from above.
the solution i posted basically works adding routes to the router, as i mentioned i used the `pages:extend` hook to prevent nuxt from creating routes based on the files situated under `~/pages`.

This led to a unwanted behavior i have realized late in the project: The resolved components from the pages directory wouldnt react on page-specific functions anymore like `definePageMeta`. i could not set the layout which led me to believe that my approach was wrong.

So what i am doing now is to let nuxt parse through the pages directory and then in a router plugin i edit the router by removing routes i don't need and adding new ones (i.e. based on an API response) 🤓

**Reply by ajoliveau** [reply]
I'm trying to use this approach (a plugin which adds routes dynamically at runtime), the routes are registered correctly if I look at the devtools, they even say it "matches" the correct route. But it always loads the error.vue page, and it warns of a "no match found" before it even registers the route.
I wonder if the plugin runs too late in the lifecycle...

![image](https://github.com/nuxt/nuxt/assets/3447717/f83b24fd-45d1-497e-967a-74d25d42c582)

**Reply by aceHubert** [reply]
the problem is router.afterEach in nuxt:router plugin execute immediately and show error, it would not clear error in next steps
